File: BOJ(Java 공부용)/11505.java
import java.io.*;
import java.util.StringTokenizer;

public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    //    static int[] gox = {0, 1, -1, 0};
    //    static int[] goy = {1, 0, 0, -1};
    static long mod = 1000000007;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken()), m = Integer.parseInt(st.nextToken()), k = Integer.parseInt(st.nextToken());
        int h = 1;
        while (h < n) h <<= 1;
        long[] seg = new long[h * 2];
        for (int i = 1; i <= n; i++) {
            st = new StringTokenizer(br.readLine());
            long value = Long.parseLong(st.nextToken());
            update(i, value, h, seg);
        }
        for (int i = 0; i < m + k; i++) {
            st = new StringTokenizer(br.readLine());
            int choice = Integer.parseInt(st.nextToken());
            if (choice == 1) {
                int b = Integer.parseInt(st.nextToken());
                long c = Long.parseLong(st.nextToken());
                update(b, c, h, seg);
            } else {
                int left = Integer.parseInt(st.nextToken()), right = Integer.parseInt(st.nextToken());
                bw.write(query(1, left, right, 1, h, seg) + "\n");
            }
        }
        bw.flush();
        bw.close();
    }

    private static long query(int nodenum, int left, int right, int nodeleft, int noderight, long[] seg) {
        if (left > noderight || nodeleft > right) return 1;
        else if (left <= nodeleft && noderight <= right) return seg[nodenum];
        int mid = (nodeleft + noderight) / 2;
        return (query(nodenum * 2, left, right, nodeleft, mid, seg) * query(nodenum * 2 + 1, left, right, mid + 1, noderight, seg)) % mod;
    }

    private static void update(int index, long value, int h, long[] seg) {
        index += h - 1;
        seg[index] = value;
        while (index > 1) {
            index /= 2;
            seg[index] = (seg[index * 2] * seg[index * 2 + 1]) % mod;
        }
    }
}

File: BOJ(Kotlin)/11505.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

val mod = 1000000007

fun main(args : Array<String>) {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))
    var st = StringTokenizer(br.readLine())
    val n = st.nextToken().toInt()
    val m = st.nextToken().toInt()
    val k = st.nextToken().toInt()
    var h = 1
    while(h<n) h *= 2
    val v = LongArray(h*2){0}

    for (i in 1 .. n) {
        val num = br.readLine().toLong()
        update(i,num,h,v)
    }

    for (i in 0 .. m+k -1) {
        st = StringTokenizer(br.readLine())
        val a = st.nextToken().toInt()
        val b = st.nextToken().toInt()
        val c : Long = st.nextToken().toLong()
        if (a == 1) {
            update(b,c,h,v)
        }
        else {
            val ans = query(1,b,c.toInt(),1,h,v)
            bw.write(ans.toString()+"\n")
        }
    }
    bw.flush()
    bw.close()
}

fun update(i : Int, value : Long, h : Int, v : LongArray) {
    var index = i
    index += h-1

    v[index] = value
    while (index > 1) {
        index /= 2
        v[index] = (v[index*2] * v[index*2+1] ) % mod
    }
}

fun query(nodeNum : Int, left : Int, right : Int, nodeLeft : Int, nodeRight: Int, v : LongArray) : Long {
    if (nodeLeft > right || nodeRight < left)  return 1
    else if (left <= nodeLeft && nodeRight <= right) return v[nodeNum]
    val mid  = (nodeLeft+nodeRight) / 2
    return (query(nodeNum * 2, left, right,nodeLeft,mid,v) *  query(nodeNum * 2 +1, left,right, mid+1,nodeRight, v)) %mod
}